### PR TITLE
[Local GC] Implement loader protocol for a standalone GC

### PR DESCRIPTION
**Pull request by swgillespie**
This PR implements the EE portion of the local GC loader protocol as specified [here](https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/standalone-gc-loading.md). There are several parts to this PR:

* This PR introduces a new file, `src\gc\gcload.cpp`, which has code pertaining to the startup path of the GC. Previously this code was in `src\gc\gccommon.cpp`, which was not a particularly good place for it. `gcload.cpp` defines the `GC_VersionInfo` and `GC_Initialize` symbols that the spec document defines as the standard entry points for the GC.

* A lot of code was moved from `InitializeGarbageCollector` in `ceemain.cpp`  to `GCHeapUtilities::LoadAndInitialize` in `gcheaputilities.cpp`. This function (and the two helper functions it references) contains the EE's implementation of the loader protocol and is in general for initializing the GC, whether or not we are using a standalone GC.

* The build option `FEATURE_STANDALONE_GC_ONLY` was removed. We are already in a place where we depend on some handle table internals that require sections of the handle table to always be linked-in to the DAC (https://github.com/dotnet/coreclr/pull/12458). The loader code that I wrote for this PR assumes that there will always be a linked-in GC, which we will load if no standalone GC was given to us.

This PR does not address the DAC portions of the loader document - that will be addressed in a separate PR.

### CODE REVIEW

**Comment by swgillespie** [comment]
It's unfortunate that we don't really have a good way to test this code using our current test harness but I have been validating each branch locally with a debugger. Perhaps it's not worth the effort but maybe it would be interesting to have a simple c++ console app that hosts coreclr and tries to load various local GCs, each with versions in such a way that it exercises different paths?

File: src/vm/gcheaputilities.cpp [hunk 1]
```diff
@@ -36,3 +37,198 @@ bool g_sw_ww_enabled_for_gc_heap = false;
...
+namespace
+{
+
+// Loads and initializes a standalone GC, given the path to the GC
+// that we should load. Returns S_OK on success and the failed HRESULT
+// on failure.
+//
+// See Documentation/design-docs/standalone-gc-loading.md for details
+// on the loading protocol in use here.
+HRESULT LoadAndInitializeGC(TCHAR* standaloneGcLocation)
```
Comment (jkotas) [inline src/vm/gcheaputilities.cpp]: `TCHAR` ? PCWSTR would be more appropriate here, I think.

File: build.cmd [hunk 2]
```diff
@@ -94,7 +94,6 @@ set __BuildTypeDebug=0
 set __BuildTypeChecked=0
 set __BuildTypeRelease=0
 set __BuildStandaloneGC="-DFEATURE_STANDALONE_GC=0"
```
Comment (jkotas) [inline build.cmd]: Do you still need this? Why not build the clrgc.dll all the time?

File: src/vm/gcheaputilities.cpp (hunk 1 above)
Comment (swgillespie) [reply src/vm/gcheaputilities.cpp]: Ok!

File: build.cmd (hunk 2 above)
Comment (swgillespie) [reply build.cmd]: I guess I just haven't flipped the switch yet (I recall this comment: https://github.com/dotnet/coreclr/pull/10951#issuecomment-294044213). I'd be happy to do that now!

Comment (swgillespie) [reply build.cmd]: Ah, I recall that `FEATURE_EVENT_TRACE` is currently disabled when building to accommodate a standalone GC. It really only needs to be disabled for `clrgc`. This is an oversight that I can probably fix pretty easily with this PR, I think - I'll give it a try.

File: src/vm/gcheaputilities.cpp [hunk 3]
```diff
@@ -36,3 +37,198 @@ bool g_sw_ww_enabled_for_gc_heap = false;
...
+    GC_LOAD_STATUS_GET_VERSIONINFO,
+    GC_LOAD_STATUS_CALL_VERSIONINFO,
+    GC_LOAD_STATUS_DONE_VERSION_CHECK,
+    GC_LOAD_STATUS_GET_INITIALIZE,
+    GC_LOAD_STATUS_LOAD_COMPLETE
+};
+
+// Load status of the GC. If GC loading fails, the value of this
+// global indicates where the failure occured.
+GC_LOAD_STATUS g_gc_load_status = GC_LOAD_STATUS_START;
```
Comment (sergiy-k) [inline src/vm/gcheaputilities.cpp]: Does it make sense to set the status to Start only when the code actually reaches the LoadAndInitializeGC function?

**Comment by sergiy-k** [comment]
LGTM

**Comment by Maoni0** [comment]
LGTM

File: src/vm/gcheaputilities.cpp (hunk 3 above)
Comment (swgillespie) [reply src/vm/gcheaputilities.cpp]: Yeah, I can do that - that way we could differentiate between a failure during the load and some other unrelated failure before the load.

**Comment by swgillespie** [comment]
@dotnet-bot test this please

**Comment by swgillespie** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Comment by swgillespie** [comment]
The most recent commit turns `FEATURE_STANDALONE_GC` on by default, which in turn makes `clrgc.dll` build by default. `FEATURE_STANDALONE_GC` increases the size of `coreclr.dll` by 0.1% but otherwise leaves normal operation of CoreCLR unaffected.

**Comment by janvorli** [comment]
@swgillespie I wonder - do we generate the version stamp into the new libclrgc.so, like we do for all other shared libraries? You can check by:
```sh
strings libclrgc.so | grep "@(#)"
```
It should print something like 
`@(#)Version 4.6.25824.0 @BuiltBy: janvorli-JANVORLI-UBUNTU2 @SrcCode: https://github.com/dotnet/coreclr/tree/15e1472aab7ed42dafde0331ed6cece465a0763d`

**Comment by swgillespie** [comment]
@janvorli Apparently it does:

```
$ strings bin/Product/OSX.x64.Debug/libclrgc.dylib | grep "@(#)"
@(#)Version 4.6.25826.0 @BuiltBy: sean-Seans-MacBook-Pro-2.local @SrcCode: https://github.com/dotnet/coreclr/tree/9ad4e5a2db7a11201bcae36de28153d8caa50269
```

I'll be honest - I'm a little surprised that worked 😆 . I suppose `add_library_clr` in CMake takes care of that?

**Comment by swgillespie** [comment]
I'll tackle the arm and arm64 issues today - just need to emit the correct compiler intrinsics for these architectures.

The x86 failure looks like cmake fail - the VM and non-standalone GC should be building with `WRITE_BARRIER_CHECK` but the standalone GC shouldn't.

**Comment by swgillespie** [comment]
I think this is ready to merge, if there aren't any other objections - I believe i've addressed all of the feedback in this thread.

**Comment by swgillespie** [comment]
thanks for the reviews!

**Pull request merged**